src/bin/test_providers.rs

```diff
@@ -0,0 +1,444 @@
+                Uniform::from(Range {
+                    start: BERLIN_BLOCK_NUMBER,
+                    end: LONDON_BLOCK_NUMBER,
+                })
+                .sample_iter(&mut rng)
+                .take(sample_size)
+                .collect(),
+                SuccessRate::default(),
+            )),
+            Ranges::London((
+                Uniform::from(Range {
+                    start: LONDON_BLOCK_NUMBER,
+                    end: MERGE_BLOCK_NUMBER,
+                })
+                .sample_iter(&mut rng)
+                .take(sample_size)
+                .collect(),
+                SuccessRate::default(),
+            )),
+            Ranges::Merge((
+                Uniform::from(Range {
+                    start: MERGE_BLOCK_NUMBER,
+                    end: SHANGHAI_BLOCK_NUMBER,
+                })
+                .sample_iter(&mut rng)
+                .take(sample_size)
+                .collect(),
+                SuccessRate::default(),
+            )),
+            Ranges::Shanghai((
+                Uniform::from(Range {
+                    start: SHANGHAI_BLOCK_NUMBER,
+                    end: latest_block,
+                })
+                .sample_iter(&mut rng)
+                .take(sample_size)
+                .collect(),
+                SuccessRate::default(),
+            )),
+        ]
+    }
+}
+
+#[derive(Debug, PartialEq, Default)]
+struct SuccessRate {
+    failures: u64,
+    total: u64,
+}
+
+#[derive(Debug, PartialEq)]
+enum Providers {
+    PandaGeth(SuccessRate),
+    PandaErigon(SuccessRate),
+    PandaArchive(SuccessRate),
+    Infura(SuccessRate),
+}
+
+impl fmt::Display for Providers {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Providers::PandaGeth(_) => write!(f, "PandaGeth"),
+            Providers::PandaErigon(_) => write!(f, "PandaErigon"),
+            Providers::PandaArchive(_) => write!(f, "PandaArchive"),
+            Providers::Infura(_) => write!(f, "Infura"),
+        }
+    }
+}
+
+impl Providers {
+    fn display_summary(&self) {
+        let success_rate = match self {
+            Providers::PandaGeth(success_rate)
+            | Providers::PandaErigon(success_rate)
+            | Providers::PandaArchive(success_rate)
+            | Providers::Infura(success_rate) => success_rate,
+        };
+        info!(
+            "Provider: {} // Failure Rate: {:?} / {:?}",
+            self, success_rate.failures, success_rate.total
+        );
+    }
+
+    fn update_success_rate(&mut self, failures: u64, total: u64) {
+        match self {
+            Providers::PandaGeth(success_rate)
+            | Providers::PandaErigon(success_rate)
+            | Providers::PandaArchive(success_rate)
+            | Providers::Infura(success_rate) => {
+                success_rate.failures += failures;
+                success_rate.total += total;
+            }
+        }
+    }
+
+    fn iterator() -> Vec<Providers> {
```

Comment:
```suggestion
    into_vec() -> Vec<Providers> {
```
We can also leverage here https://docs.rs/strum_macros/0.25.3/strum_macros/derive.EnumIter.html here but I'm not sure if it is worth importing `strum` only for this use case.